source/common/formatter/substitution_formatter.h

```diff
@@ -18,6 +18,19 @@
 namespace Envoy {
 namespace Formatter {
 
+class CommandSyntaxChecker {
+public:
+  using CommandSyntaxFlags = uint32_t;
+  static constexpr CommandSyntaxFlags COMMAND_ONLY = 0;
+  static constexpr CommandSyntaxFlags PARAMS_REQUIRED = 1 << 0;
+  static constexpr CommandSyntaxFlags PARAMS_OPTIONAL = 1 << 1;
+  static constexpr CommandSyntaxFlags LENGTH_ALLOWED = 1 << 2;
+
+  static void VerifySyntax(CommandSyntaxChecker::CommandSyntaxFlags flags,
+                           const std::string& command, const std::string& subcommand,
+                           const absl::optional<size_t>& length);
```

Comment:
Random passer-by comment: the way that CEL deals with parsing is that it:
1) has an alternate AST form in proto so that the parsing is not done on the data plane;
2) uses ANTLR for a strict grammar.

Do you think we can use those techniques here to improve security of the format strings?

Reply:
The problem here is readability. Security is not so much since this is parsed during control plane updates. I would go for replacing it with ANTLR generated parser, since working with the grammar file is much easier.

Reply:
Thanks for feedback. I will experiment/prototype with ANTLR.